juju/juju
# uniter: Be smarter about when we run config-changed (part 1)

## Description of change
At the moment, when a controller agent restarts all of the unit agents that were connected to it reconnect and then immediately fire config-changed, even if no configuration has actually changed. On controllers with thousands of units, this can overload the controller agent while all the units get all their config and run any other hook commands in response.

The uniter always runs config-changed when it first starts because it doesn't track any local state that would enable it to see that the config is the same and avoid running the hook.

Change the API to expose Uniter.WatchConfigSettingsHash - this works like .WatchConfigSettings, except it sends a hash of the current config when it signals. The uniter stores that value in the local uniter state file, so that it persists across agent restarts, and checks it before firing config-changed.

This bumps the uniter API to version 9.

This PR plumbs the config hash through the API and uniter, but the uniter also watches trust config and addresses, so the behaviour won't change until those other parts are done in followups.

## QA steps

* Bootstrap a controller and model, deploy some applications and relate them. 
* See that the uniter state files for various units includes the config-hash.
* Change a config value for one of the applications.
* See that the config-hash changes when the config-changed hook fires.

## Documentation changes

None
